File: scratch/search_transcripts.py
#!/usr/bin/env python3
"""
search_transcripts.py

Instant Deep-Link Timestamp Searcher CLI across all 1,740 YAML transcripts.
Uses the `timestamps: [[char_offset, seconds], ...]` index to binary-search
the exact playback second for any keyword or phrase match and produces
direct YouTube jump links (`https://youtu.be/{id}?t={seconds}s`).

Usage:
  python3 search_transcripts.py "borrow checker"
  python3 search_transcripts.py "docker" --type short
  python3 search_transcripts.py "dopamine" --channel "HealthyGamerGG"
  python3 search_transcripts.py "pricing" --domain "business_systems"
"""

import os
import sys
import re
import bisect
import argparse
import yaml

try:
    from yaml import CSafeLoader as Loader
except ImportError:
    from yaml import SafeLoader as Loader

FORMATTED_DIR = "formatted_transcripts"
TAXONOMY_FILE = "taxonomy/video_metadata_registry.yaml"

# ANSI color codes
CYAN = "\033[96m"
GREEN = "\033[92m"
YELLOW = "\033[93m"
MAGENTA = "\033[95m"
BOLD = "\033[1m"
DIM = "\033[2m"
RESET = "\033[0m"


def format_seconds(seconds):
    hh = seconds // 3600
    mm = (seconds % 3600) // 60
    ss = seconds % 60
    if hh > 0:
        return f"{hh:02d}:{mm:02d}:{ss:02d}"
    return f"{mm:02d}:{ss:02d}"


def build_youtube_jump_url(base_url, seconds, video_id=""):
    if not base_url and video_id:
        base_url = f"https://www.youtube.com/watch?v={video_id}"
    if not base_url:
        return ""
    # Strip any existing timestamp param
    clean_url = re.sub(r'[?&]t=\d+s?', '', base_url)
    sep = "&" if "?" in clean_url else "?"
    return f"{clean_url}{sep}t={seconds}s"


def find_timestamp_for_char(timestamps, char_index, text=""):
    """
    Given a list of [char_pos, seconds], finds the cue matching char_index
    via binary search. If timestamps list is empty, falls back to scanning
    inline timestamp tags like [26.64s] or 01:23 in the preceding text.
    """
    if timestamps:
        keys = [item[0] for item in timestamps]
        idx = bisect.bisect_right(keys, char_index) - 1
        if idx >= 0:
            return timestamps[idx][1]
    
    # Fallback to inline timestamp markers if present in raw transcript
    if text:
        prefix = text[:char_index]
        inline_matches = list(re.finditer(r'\[(\d+(?:\.\d+)?)s?\]', prefix))
        if inline_matches:
            return int(round(float(inline_matches[-1].group(1))))
        ts_matches = list(re.finditer(r'\b(\d{1,2}):(\d{2})(?::(\d{2}))?\b', prefix))
        if ts_matches:
            m = ts_matches[-1]
            if m.group(3):
                return int(m.group(1)) * 3600 + int(m.group(2)) * 60 + int(m.group(3))
            return int(m.group(1)) * 60 + int(m.group(2))

    return 0


def extract_snippet(text, match_start, match_end, radius=90):
    start = max(0, match_start - radius)
    end = min(len(text), match_end + radius)
    
    prefix = ("..." if start > 0 else "") + text[start:match_start]
    match = text[match_start:match_end]
    suffix = text[match_end:end] + ("..." if end < len(text) else "")
    
    # Clean newlines for terminal display
    prefix = prefix.replace('\n', ' ')
    match = match.replace('\n', ' ')
    suffix = suffix.replace('\n', ' ')
    
    return f"{prefix}{BOLD}{YELLOW}{match}{RESET}{suffix}"


def main():
    parser = argparse.ArgumentParser(
        description="Search across 1,740 YouTube transcripts with instant deep-link timestamps."
    )
    parser.add_argument("query", help="Keyword or regex phrase to search for")
    parser.add_argument("--channel", help="Filter by channel name (case-insensitive substring)")
    parser.add_argument("--domain", help="Filter by domain (e.g. software_engineering, business_systems)")
    parser.add_argument("--type", choices=["short", "video"], help="Filter by content type")
    parser.add_argument("--limit", type=int, default=10, help="Maximum results to return (default: 10)")
    parser.add_argument("--exact", action="store_true", help="Match exact word boundaries")
    parser.add_argument("--regex", action="store_true", help="Treat query as regular expression")

    args = parser.parse_args()

    # Load video_metadata_registry if domain filter is specified
    domain_map = {}
    if args.domain:
        if os.path.exists(TAXONOMY_FILE):
            with open(TAXONOMY_FILE, 'r', encoding='utf-8') as tf:
                reg = yaml.load(tf, Loader=Loader)
                for vid, meta in reg.items():
                    domain_map[vid] = meta.get('domain', '')

    if not os.path.exists(FORMATTED_DIR):
        print(f"Error: Directory '{FORMATTED_DIR}' not found.")
        sys.exit(1)

    # Compile search pattern
    flags = re.IGNORECASE
    if args.regex:
        pattern_str = args.query
    elif args.exact:
        pattern_str = r'\b' + re.escape(args.query) + r'\b'
    else:
        pattern_str = re.escape(args.query)

    try:
        search_regex = re.compile(pattern_str, flags)
    except re.error as e:
        print(f"Invalid regex: {e}")
        sys.exit(1)

    files = sorted([f for f in os.listdir(FORMATTED_DIR) if f.endswith('.yaml')])
    results = []
    scanned_count = 0

    print(f"\n🔍 Searching for '{args.query}' across {len(files)} formatted transcripts...")

    for f in files:
        fpath = os.path.join(FORMATTED_DIR, f)
        with open(fpath, 'r', encoding='utf-8') as fp:
            doc = yaml.load(fp, Loader=Loader)

        scanned_count += 1
        v_type = doc.get('type', 'video')
        channel = doc.get('channel', 'Unknown')
        video_id = doc.get('video_id', '')

        # Apply filters
        if args.type and v_type != args.type:
            continue
        if args.channel and args.channel.lower() not in channel.lower():
            continue
        if args.domain and domain_map.get(video_id) != args.domain:
            continue

        transcript = doc.get('transcript', '')
        if not transcript:
            continue

        match = search_regex.search(transcript)
        if match:
            timestamps = doc.get('timestamps', [])
            matched_char_start = match.start()
            matched_char_end = match.end()

            seconds = find_timestamp_for_char(timestamps, matched_char_start, transcript)
            time_str = format_seconds(seconds)
            jump_url = build_youtube_jump_url(doc.get('url', ''), seconds, video_id)
            snippet = extract_snippet(transcript, matched_char_start, matched_char_end)

            results.append({
                'title': doc.get('title', 'Untitled'),
                'channel': channel,
                'type': v_type,
                'duration': doc.get('duration', '00:00'),
                'time_str': time_str,
                'seconds': seconds,
                'jump_url': jump_url,
                'snippet': snippet,
                'video_id': video_id
            })

            if len(results) >= args.limit:
                break

    print(f"🎯 Found {len(results)} matches (scanned {scanned_count} files):\n")

    if not results:
        print("   No matching transcripts found. Try broadening your query or removing filters.")
        return

    for i, res in enumerate(results, 1):
        type_badge = f"{MAGENTA}[SHORT]{RESET}" if res['type'] == 'short' else f"{CYAN}[VIDEO]{RESET}"
        print(f"{BOLD}{i}. {res['title']}{RESET} {type_badge}")
        print(f"   {DIM}Channel:{RESET} {res['channel']}  |  {DIM}Duration:{RESET} {res['duration']}  |  {DIM}Timestamp:{RESET} {GREEN}{res['time_str']}{RESET} ({res['seconds']}s)")
        print(f"   {DIM}Context:{RESET} \"{res['snippet']}\"")
        if res['jump_url']:
            print(f"   👉 {BOLD}{CYAN}Jump to video:{RESET} {res['jump_url']}")
        print()


if __name__ == '__main__':
    main()

File: src/bin/search_transcripts.rs
use rayon::prelude::*;
use regex::Regex;
use serde::Deserialize;
use std::env;
use std::fs;
use std::path::{Path, PathBuf};

#[derive(Debug, Deserialize)]
struct YamlTranscript {
    #[serde(default)]
    video_id: String,
    #[serde(default)]
    title: String,
    #[serde(default)]
    channel: String,
    #[serde(default)]
    url: String,
    #[serde(default)]
    duration: String,
    #[serde(default, rename = "type")]
    content_type: String,
    #[serde(default)]
    transcript: String,
    #[serde(default)]
    timestamps: Vec<[usize; 2]>,
}

struct SearchMatch {
    title: String,
    channel: String,
    duration: String,
    content_type: String,
    time_str: String,
    seconds: usize,
    jump_url: String,
    snippet: String,
}

fn format_seconds(seconds: usize) -> String {
    let hh = seconds / 3600;
    let mm = (seconds % 3600) / 60;
    let ss = seconds % 60;
    if hh > 0 {
        format!("{:02}:{:02}:{:02}", hh, mm, ss)
    } else {
        format!("{:02}:{:02}", mm, ss)
    }
}

fn find_timestamp_for_char(timestamps: &[[usize; 2]], char_idx: usize) -> usize {
    if timestamps.is_empty() {
        return 0;
    }
    let idx = timestamps.partition_point(|p| p[0] <= char_idx);
    if idx > 0 {
        timestamps[idx - 1][1]
    } else {
        timestamps[0][1]
    }
}

fn extract_snippet(text: &str, m_start: usize, m_end: usize, radius: usize) -> String {
    let s = m_start.saturating_sub(radius);
    let e = (m_end + radius).min(text.len());

    let prefix = if s > 0 { "..." } else { "" };
    let prefix_text = &text[s..m_start].replace('\n', " ");
    let matched_text = &text[m_start..m_end].replace('\n', " ");
    let suffix_text = &text[m_end..e].replace('\n', " ");
    let suffix = if e < text.len() { "..." } else { "" };

    format!("{}{}\x1b[1;33m{}\x1b[0m{}{}", prefix, prefix_text, matched_text, suffix_text, suffix)
}

fn print_usage() {
    println!("Usage: search_transcripts <query> [options]");
    println!("Options:");
    println!("  --channel <name>      Filter by channel name");
    println!("  --type <video|short>  Filter by content type");
    println!("  --limit <number>      Maximum matches to show (default: 10)");
    println!("  --exact               Match exact word boundaries");
    println!("  --regex               Treat query as regex");
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        print_usage();
        return Ok(());
    }

    let mut query = String::new();
    let mut channel_filter: Option<String> = None;
    let mut type_filter: Option<String> = None;
    let mut limit = 10usize;
    let mut exact = false;
    let mut is_regex = false;

    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "-h" | "--help" => {
                print_usage();
                return Ok(());
            }
            "--channel" => {
                i += 1;
                if i < args.len() {
                    channel_filter = Some(args[i].to_lowercase());
                }
            }
            "--type" => {
                i += 1;
                if i < args.len() {
                    type_filter = Some(args[i].to_lowercase());
                }
            }
            "--limit" => {
                i += 1;
                if i < args.len() {
                    if let Ok(l) = args[i].parse::<usize>() {
                        limit = l;
                    }
                }
            }
            "--exact" => {
                exact = true;
            }
            "--regex" => {
                is_regex = true;
            }
            other => {
                if query.is_empty() {
                    query = other.to_string();
                }
            }
        }
        i += 1;
    }

    if query.is_empty() {
        eprintln!("Error: No search query provided.");
        print_usage();
        std::process::exit(1);
    }

    let formatted_dir = Path::new("formatted_transcripts");
    if !formatted_dir.exists() {
        eprintln!("Error: Directory 'formatted_transcripts' not found.");
        std::process::exit(1);
    }

    let pattern_str = if is_regex {
        query.clone()
    } else if exact {
        format!(r"\b{}\b", regex::escape(&query))
    } else {
        regex::escape(&query)
    };

    let search_rx = match Regex::new(&format!("(?i){}", pattern_str)) {
        Ok(r) => r,
        Err(e) => {
            eprintln!("Invalid search pattern: {}", e);
            std::process::exit(1);
        }
    };

    let mut entries: Vec<PathBuf> = fs::read_dir(formatted_dir)?
        .filter_map(|res| res.ok())
        .map(|entry| entry.path())
        .filter(|p| p.extension().map_or(false, |ext| ext == "yaml"))
        .collect();
    entries.sort();

    println!(
        "\n\x1b[36m🔍 Searching for '{}' across {} formatted transcripts...\x1b[0m",
        query, entries.len()
    );

    let matches: Vec<SearchMatch> = entries
        .par_iter()
        .filter_map(|path| {
            let content = fs::read_to_string(path).ok()?;
            let doc: YamlTranscript = serde_yaml::from_str(&content).ok()?;

            if let Some(ref cf) = channel_filter {
                if !doc.channel.to_lowercase().contains(cf) {
                    return None;
                }
            }
            if let Some(ref tf) = type_filter {
                if !doc.content_type.to_lowercase().contains(tf) {
                    return None;
                }
            }

            if doc.transcript.is_empty() {
                return None;
            }

            if let Some(m) = search_rx.find(&doc.transcript) {
                let m_start = m.start();
                let m_end = m.end();
                let seconds = find_timestamp_for_char(&doc.timestamps, m_start);
                let time_str = format_seconds(seconds);
                let clean_url = doc.url.split('?').next().unwrap_or(&doc.url);
                let jump_url = format!("{}?v={}&t={}s", clean_url, doc.video_id, seconds);
                let snippet = extract_snippet(&doc.transcript, m_start, m_end, 85);

                Some(SearchMatch {
                    title: doc.title,
                    channel: doc.channel,
                    duration: doc.duration,
                    content_type: doc.content_type,
                    time_str,
                    seconds,
                    jump_url,
                    snippet,
                })
            } else {
                None
            }
        })
        .collect();

    println!(
        "\x1b[32m🎯 Found {} matches (showing up to {}):\x1b[0m\n",
        matches.len(),
        limit
    );

    if matches.is_empty() {
        println!("   No matching transcripts found. Try broadening your query or removing filters.");
        return Ok(());
    }

    for (idx, m) in matches.iter().take(limit).enumerate() {
        let type_badge = if m.content_type == "short" {
            "\x1b[1;35m[SHORT]\x1b[0m"
        } else {
            "\x1b[1;36m[VIDEO]\x1b[0m"
        };

        println!("{}. \x1b[1m{}\x1b[0m {}", idx + 1, m.title, type_badge);
        println!(
            "   \x1b[2mChannel:\x1b[0m {}  |  \x1b[2mDuration:\x1b[0m {}  |  \x1b[2mTimestamp:\x1b[0m \x1b[1;32m{}\x1b[0m ({}s)",
            m.channel, m.duration, m.time_str, m.seconds
        );
        println!("   \x1b[2mContext:\x1b[0m \"{}\"", m.snippet);
        if !m.jump_url.is_empty() {
            println!("   👉 \x1b[1;36mJump to video:\x1b[0m \x1b[4m{}\x1b[0m", m.jump_url);
        }
        println!();
    }

    Ok(())
}
